lib/SILOptimizer/Mandatory/AccessEnforcementSelection.cpp

```diff
@@ -0,0 +1,334 @@
+    auto &succState = StateMap[succ];
+
+    // If updating the successor changes it in a way that will
+    // require us to update its successors, add it to the worklist.
+    if (succState.adjustForEscapeInPredecessor()) {
+      if (!succState.IsInWorklist) {
+        succState.IsInWorklist = true;
+        Worklist.push_back(succ);
+      }
+    }
+  }
+}
+
+bool SelectEnforcement::hasPotentiallyEscapedAt(SILInstruction *point) {
+  auto bb = point->getParent();
+
+  // If we're not tracking anything for the whole block containing
+  // the instruction, we're done; it hasn't escaped here.
+  auto it = StateMap.find(bb);
+  if (it == StateMap.end())
+    return false;
+
+  // If the tracking information says there are escapes before entry,
+  // we're done; it has potentially escaped.
+  const auto &state = it->second;
+  if (state.HasPotentiallyEscapedAtEntry)
+    return true;
+
+  // Okay, there must be an escape within this block.
+  assert(state.HasEscape);
+  for (auto ii = point->getIterator(), ie = bb->begin(); ii != ie; ) {
+    auto inst = &*--ii;
+
+    // Maybe just record the first escape in the block and see if we
+    // come after it?
+    if (Escapes.count(inst))
+      return true;
+  }
+
+  return false;
+}
+
+bool SelectEnforcement::hasPotentiallyEscapedAtAnyReachableBlock(
+                                                      BeginAccessInst *access) {
+  // Fast path: we're not tracking any escapes.  (But the box should
+  // probably have been promoted to the stack in this case.)
+  if (StateMap.empty())
+    return false;
+
+  auto bb = access->getParent();
+
+  assert(Worklist.empty());
+  SmallPtrSet<SILBasicBlock*, 8> visited;
+  visited.insert(bb);
+  Worklist.push_back(bb);
+
+  while (!Worklist.empty()) {
+    bb = Worklist.pop_back_val();
+    assert(visited.count(bb));
+
+    // If we're tracking information for this block, there's an escape.
+    if (StateMap.count(bb))
+      return true;
+
+    // Add all reachable successors.
+    for (SILBasicBlock *succ : bb->getSuccessors()) {
+      if (visited.insert(succ).second)
+        Worklist.push_back(succ);
+    }
+  }
+
+  // No reachable block has an escape.
+  return false;
+}
+
+void SelectEnforcement::updateAccesses() {
+  for (auto access : Accesses) {
+    updateAccess(access);
+  }
+}
+
+void SelectEnforcement::updateAccess(BeginAccessInst *access) {
+  assert(access->getEnforcement() == SILAccessEnforcement::Unknown);
+
+  // Check whether the variable escaped before any of the end_accesses.
+  bool anyDynamic = false;
+  bool hasEndAccess = false;
+  for (auto endAccess : access->getEndAccesses()) {
+    hasEndAccess = true;
+    if (hasPotentiallyEscapedAt(endAccess)) {
+      anyDynamic = true;
+      break;
+    }
+  }
+
+  // If so, make the access dynamic.
+  if (anyDynamic)
+    return setDynamicEnforcement(access);
+
+  // Otherwise, if there are no end_access instructions,
+  // check the terminators of every reachable block.
```

Comment:
I'm not sure what "check the terminators" means" This code just checks if there are any uses reachable from this begin_access.